### PR TITLE
Use simdutf8 for UTF-8 validation

### PR DESCRIPTION
**Pull request by hkratz**
This went rather well. LLVM does an awesome job optimizing and the benchmark results are virtually identical.

Still experimenting a bit with the API in https://github.com/rusticstuff/simdutf8/pull/44.

Comments welcome.

### CODE REVIEW

**Comment by hkratz** [comment]
The remaining aarch64 build failure is stdarch's fault, should be fixed soon with https://github.com/rust-lang/stdarch/pull/1157.

File: src/lib.rs [hunk 1]
```diff
@@ -190,6 +195,12 @@ pub use crate::sse42::deser::*;
     target_feature = "neon"
 ))))]
 use crate::sse42::stage1::{SimdInput, SIMDINPUT_LENGTH, SIMDJSON_PADDING};
+#[cfg(all(not(any(
+    target_feature = "sse4.2",
+    target_feature = "avx2",
+    target_feature = "neon"
+))))]
```
Comment (Licenser) [inline src/lib.rs:202]: I think this guard is slighty off
```suggestion
#[cfg(all(
    target_feature = "sse4.2",
    not(any(
      target_feature = "avx2",
      target_feature = "neon"
))))]
```

**Comment by Licenser** [comment]
Heya, I really  love this :D it's great to reduce the service area and turn the utf8 checking in a standalone crate that can be used outside of simd_json! Also makes it a lot easier to iterate and improve on them separately. Absolutely great!

I'll run it through a few benchmarks just to be sure but I see nothing stopping this 👍 thank you so much

File: src/lib.rs (hunk 1 above)
Comment (hkratz) [reply src/lib.rs:202]: I copied that from the preceding lines. I assume that is meant to be the fallback case?

Comment (Licenser) [reply src/lib.rs:202]: My apologies you're right :) I didn't unfold the previous lines!

**Comment by hkratz** [comment]
I am very excited about this as well! I think it is awesome, that it works with a properly abstracted API without performance loss across the crate boundary. LLVM is able to remove the 64 byte load from memory done in simdutf8 and just use the registers it has already loaded the data into so that no performance is lost.

BTW: Tarpaulin runs fine for me locally not sure what the problem with the CI is.

**Comment by Licenser** [comment]
The error doesn't seem to be realated to your code, it was a earlier PR that for some reason didn't trigger the CI, I'm reverting it then a rebase should fix the issues 👍

**Comment by Licenser** [comment]
rebasing should do the trick 👍

**Comment by Licenser** [comment]
The tarpaulin failures are OK, they just fail because a PR does not have permissions to upload to codecov. the aarch64  are known failing as you mentioned :) so neither of those two are a problem

**Review by Licenser — APPROVED** [review]
Great work! 🚀

**Comment by Licenser** [comment]
I'm good with emrging this, let me know when you're ready and mark it as non draft then well bring it to main :)

**Comment by hkratz** [comment]
Sounds great! I will add some unit tests and documentation to the simdutf8 streaming validation API, make a release and mark the pull request as ready when that is done.

**Comment by hkratz** [comment]
There were some internal GitHub errors in the last CI so I triggered another one.

**Comment by hkratz** [comment]
Benchmarks look good, too:
```
group                                         main                                    with-simdutf8
-----                                         ----                                    -------------
apache_builds/simd_json::to_borrowed_value    1.00    207.1±2.63µs   606.3 MB/sec     1.01   208.5±13.77µs   602.3 MB/sec
apache_builds/simd_json::to_owned_value       1.01    498.2±5.74µs   252.1 MB/sec     1.00    493.9±4.98µs   254.3 MB/sec
apache_builds/simd_json::to_tape              1.05     63.8±6.67µs  1969.9 MB/sec     1.00     60.9±3.49µs     2.0 GB/sec
canada/simd_json::to_borrowed_value           1.00      8.8±0.19ms   244.8 MB/sec     1.00      8.8±0.26ms   244.6 MB/sec
canada/simd_json::to_owned_value              1.00      8.5±0.19ms   252.6 MB/sec     1.01      8.6±0.18ms   248.9 MB/sec
canada/simd_json::to_tape                     1.00      3.8±0.18ms   564.6 MB/sec     1.00      3.8±0.14ms   566.4 MB/sec
citm_catalog/simd_json::to_borrowed_value     1.00      3.4±0.08ms   503.0 MB/sec     1.01      3.4±0.09ms   499.6 MB/sec
citm_catalog/simd_json::to_owned_value        1.01      4.9±0.14ms   349.1 MB/sec     1.00      4.8±0.12ms   352.1 MB/sec
citm_catalog/simd_json::to_tape               1.00  1530.1±130.84µs  1108.0 MB/sec    1.02  1559.8±207.04µs  1086.9 MB/sec
log/simd_json::to_borrowed_value              1.00      2.4±0.03µs   857.8 MB/sec     1.01      2.4±0.22µs   847.5 MB/sec
log/simd_json::to_owned_value                 1.03      7.8±0.07µs   265.3 MB/sec     1.00      7.6±0.24µs   273.4 MB/sec
log/simd_json::to_tape                        1.06  1365.9±35.14ns  1519.3 MB/sec     1.00  1284.5±32.78ns  1615.5 MB/sec
twitter/simd_json::to_borrowed_value          1.00   991.5±48.19µs   622.3 MB/sec     1.00   989.1±33.33µs   623.8 MB/sec
twitter/simd_json::to_owned_value             1.03      2.1±0.05ms   295.6 MB/sec     1.00      2.0±0.10ms   305.1 MB/sec
twitter/simd_json::to_tape                    1.00   453.0±18.28µs  1362.0 MB/sec     1.00   454.7±29.00µs  1356.9 MB/sec
```
So from my side this is ready for merging.

**Comment by Licenser** [comment]
Yap ran it too looks great :+1: thank you so much!

**Pull request merged**